src/datachain/lib/udf.py

```diff
@@ -300,24 +337,29 @@ def run(
         download_cb: Callback = DEFAULT_CALLBACK,
         processed_cb: Callback = DEFAULT_CALLBACK,
     ) -> Iterator[Iterable[UDFResult]]:
-        self.catalog = catalog
         self.setup()
-        prepared_inputs: abc.Generator[Sequence[Any], None, None] = (
-            self._prepare_row_and_id(row, udf_fields, cache, download_cb)
-            for row in udf_inputs
-        )
-        if self.prefetch > 0:
-            prepared_inputs = AsyncMapper(
-                _prefetch_input, prepared_inputs, workers=self.prefetch
-            ).iterate()
-
-        with contextlib.closing(prepared_inputs):
-            for id_, *udf_args in prepared_inputs:
-                result_objs = self.process_safe(udf_args)
-                udf_output = self._flatten_row(result_objs)
-                output = [{"sys__id": id_} | dict(zip(self.signal_names, udf_output))]
-                processed_cb.relative_update(1)
-                yield output
+
+        def _prepare_rows(udf_inputs) -> "abc.Generator[Sequence[Any], None, None]":
+            with safe_closing(udf_inputs):
+                for row in udf_inputs:
+                    yield self._prepare_row_and_id(
+                        row, udf_fields, catalog, cache, download_cb
+                    )
```

Comment:
Using `safe_closing` here and below so that the `udf_inputs` gets properly closed when this `_prepare_rows` gets `.close()`d.

We were using generator expression before:
```python
     prepared_inputs: abc.Generator[Sequence[Any], None, None] = (
            self._prepare_row_and_id(row, udf_fields, cache, download_cb)
            for row in udf_inputs
        )
```

Generator expression/function does not send `close` or `throw` to the subgenerator (`udf_inputs` in this case), so we have to add this contextmanager and change it to a generator function from a generator expression.

Note that this function runs in another thread in case of prefetching. So, we want `.close()` for `udf_inputs` to run in the same thread, not in a different thread when it gets garbage-collected.